[MVC - Step 2] 에단 미션 제출합니다.

Laterality opened the pull request.

Laterality committed:
Refactor component scanner

Laterality committed:
Refactor manual handler mapping adapter

Laterality committed:
Implement json view

Laterality committed:
implement user api controller

Laterality committed:
Refactor: removed legacy controller mapping

phs1116 commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/ModelAndView.java:
```diff
@@ -9,11 +9,17 @@
 
     public void renderView(HttpServletRequest req, HttpServletResponse resp) throws Exception {
+        if (view == null) {
+            view = new TextView("");
```
이런 메서드는 static factory method로도 표현 가능할 것 같네요. :)
그리고 생성시 별도 세팅을 하자 않았을 시에 Default로 해당 View를 세팅하면 이런 if문은 필요 없지 않을까요? :)

phs1116 commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/JsonView.java:
```diff
@@ -1,12 +1,41 @@
+            writer.print(tryWriteValueAsString(model.values().toArray()[0]));
+            return;
+        }
+        writer.print(tryWriteValueAsString(model));
```
너무 구현 잘하셨는데 이부분을 조금 더 간결하게도 리팩토링 할 수 있을것 같아요! :)

Laterality replied:
write하는 조건과 값만 바뀐다는 점에 착안해서 리팩터링해봤습니다!

phs1116 commented on slipp/src/main/java/slipp/controller/UserApiController.java:
```diff
@@ -0,0 +1,64 @@
+
+    private <T> T parseBody(HttpServletRequest req, Class<T> clazz) {
+        try {
+            return JsonUtils.toObject(req.getReader().lines().collect(Collectors.joining()), clazz);
```
사소한 내용이지만 ObjectMapper는 굳이 String 형태로 만들지 않아도 inputStream으로 매핑하는 기능을 제공하고 있어요.
`public <T> T readValue(InputStream src, Class<T> valueType)`
HttpServletRequest은 `request.getInputStream()` 을 제공하고요. :)

phs1116 commented on slipp/src/main/java/slipp/controller/UserApiController.java:
```diff
@@ -0,0 +1,64 @@
+    private <T> T parseBody(HttpServletRequest req, Class<T> clazz) {
+        try {
+            return JsonUtils.toObject(req.getReader().lines().collect(Collectors.joining()), clazz);
+        } catch (IOException e) {
```
Util형태로 제공할거라면 매번 사용하는 측에서 unchecked로 변환하는 것 보다는 이조차도 유틸에서 감싸주는게 좋을것 같아요. :)

Laterality replied:
위에서 피드백해주신대로 `JsonUtils.toObject`가 `InputStream`을 인자로 받을 수 있게 수정해봤는데요. `HttpServletRequest.getInputStream`이 `IOException`을 던지고 있어서 현재 위치에서 catch해야 할 것 같습니다!

phs1116 commented on slipp/src/main/java/slipp/controller/UserApiController.java:
```diff
@@ -0,0 +1,64 @@
+import java.util.stream.Collectors;
+
+@Controller
+public class UserApiController {
```
새로운 컨트롤러들(JsonView)에 대한 인수 테스트가 없는 것 같아요.  :)

phs1116 commented on slipp/src/main/java/slipp/controller/UserController.java:
```diff
@@ -2,23 +2,65 @@
+        String userId = req.getParameter("userId");
+        User user = DataBase.findUserById(userId);
+        if (user == null) {
+            throw new NullPointerException("사용자를 찾을 수 없습니다.");
```
NullPointerException가 해당 예외에 적합할까요?
프로그래밍 예외에 대해서는 표준 예외가 적합하지만,
비즈니스에 대한 예외면 비즈니스 예외를 사용하는게 적합하지 않을까 싶어요. :)
사용자가 없는 경우에 대해서 별도 예외를 처리해야할 때 NullPointerException를 핸들링 할 수도 없다는 문제도 있구요. :)

Laterality replied:
별도 예외를 사용하도록 수정했습니다!

phs1116 commented on slipp/src/main/java/slipp/controller/UserController.java:
```diff
@@ -2,23 +2,65 @@
+    @RequestMapping(value = "/users/profile", method = RequestMethod.GET)
+    public ModelAndView profile(HttpServletRequest req, HttpServletResponse resp) {
+        String userId = req.getParameter("userId");
+        User user = DataBase.findUserById(userId);
```
findUserById에 대해서 Optional을 사용하도록도 리팩토링 할 수 있겠네요. :)

phs1116 commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/RequestMappingScanner.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class RequestMappingScanner {
```
이 객체에 RequestMapping에 대해서만 책임을 지게 하고 싶었다면 instances 정보들과 instantiate는 `ComponentScanner`가 가져가고, 여기에 위임하도록 하는게 맞지 않을까요? :)

Laterality replied:
`ComponentScanner`는 스캐닝을 위한 전략(Reflection API 혹은 Reflections 외부 의존성)만을 담당하도록 하고 싶었습니다. 고민 끝에 인스턴스 관리를 위한 별도의 일급 컬렉션으로 분리해봤습니다!

phs1116 commented on slipp/src/main/java/slipp/controller/UserApiController.java:
```diff
@@ -0,0 +1,64 @@
+        User user = DataBase.findUserById(userId);
+        user.update(new User(userId, userUpdatedDto.getPassword(), userUpdatedDto.getName(), userUpdatedDto.getEmail()));
+
+        resp.setStatus(HttpServletResponse.SC_OK);
```
200 OK의 케이스에 대해서도 매번 status 선언이 필요할까요? :)

Laterality replied:
https://docs.oracle.com/cd/E23095_01/Platform.93/ATGProgGuide/html/s0803httpservletresponse01.html
기본값이 200으로 세팅되는군요!

phs1116 commented on slipp/src/main/java/slipp/controller/LoginController.java:
```diff
@@ -1,30 +1,44 @@
             req.setAttribute("loginFailed", true);
-            return "/user/login.jsp";
+            return new ModelAndView(new JspView("/user/login.jsp"));
         }
```
사소한 내용이지만 브랜치가 복잡한데 좀 더 간결하게 풀 수 있어보여요 :)

- 코드 중복
<pre>
<code>
            req.setAttribute("loginFailed", true);
            return "/user/login.jsp";
            return new ModelAndView(new JspView("/user/login.jsp"));
</code>
</pre>

Laterality replied:
`DataBase.findById`를 리팩터링하면서 함께 수정해봤습니다!

phs1116 commented on slipp/src/main/java/slipp/controller/UserController.java:
```diff
@@ -2,23 +2,65 @@
+    public ModelAndView updateUser(HttpServletRequest req, HttpServletResponse resp) {
+        User user = DataBase.findUserById(req.getParameter("userId"));
+        if (!UserSessionUtils.isSameUser(req.getSession(), user)) {
+            throw new IllegalStateException("다른 사용자의 정보를 수정할 수 없습니다.");
```
`NullPointerException`의 피드백과 마찬가지입니다. :)

phs1116 reviewed (changes requested):
안녕하세요! 에단.
이전 단계 피드백 반영 👍 
전체 컨트롤려를 신규 기능 기반으로 변경 등 
매우 깔끔하게 잘 구현해주셨어요.
다만 신규 컨트롤러등에 대해서 인수테스트가 추가되면 좋을 것 같아
관련 피드백과, 리팩토링 관련해서 피드백 추가했습니다.
참고 부탁드리고 궁금하신 점 있으시면 언제든지 연락주세요. :)

Laterality committed:
Refactor view renderer

Laterality committed:
Refactor Json utility to use with Input stream

Laterality committed:
Refactor user repository to return Optional

Laterality committed:
Refactor

- Restructured package.
- Changed to concrete exceptions in controller.
- Segregated Instance managing from request mapping scanner.
- Removed unnecessary excecption.

Laterality committed:
Add support for handlebars view

Laterality committed:
Refactor: restructure package

Laterality committed:
Refactor: restructure package

Laterality committed:
Refactor test package structure

Laterality committed:
Refactor slipp package structure

Laterality committed:
Merge branch 'feat/handlebars-support' into step-2

# Conflicts:
#	nextstep-mvc/src/main/java/nextstep/mvc/view/HandlebarsView.java

phs1116 reviewed (approved):
안녕하세요! 에단.
너무 구현 잘해주셨어요.
이번 단계 머지할게요. 
개발하시느라 수고 많으셨어요. :)
